### DISCUSSION TITLE
Why do we return void for initializers?

### DISCUSSION THREAD

**Discussion by aaronmallen** [ruby/rbs#2028] [Q&A] [answered]
Hello I was curious why the documentation repeatedly returns `void` for initializers instead of `instance`

**Comment by ParadoxV5** [comment, marked as answer]
When you instantiate a class, you call the *class* method `new` (e.g., `User.new`).
`Class#new` is the one that returns `instance`.

The initializer we typically override is the *instance* method `initialize`, which is merely a callback `Class#new` triggers.
`initialize`s seldom explicitly return `self` – think about all the `def initialize` blocks you’ve written.

> `#initialize` is often overridden, but specifying a return type can be inconvenient for type checking.
> ⸺ #2008

**Comment by ksss** [comment]
One reason is that `initialize` is a private method, and code should not depend on the return value of private methods.

---

Another reason concerns issues when defining the `initialize` method in a subclass.

For example, before the change in https://github.com/ruby/rbs/pull/2008, the definition of `BasicObject` was as follows:

```rbs
# rbs
class BasicObject
  def initialize: () -> nil
end
```

Now, suppose you define a class that inherits from `BasicObject`:

```rb
# rb
class Foo < BasicObject
  def initialize
    @foo = :default
  end
end
```

This code results in a type error because the return types of `initialize` (`nil` vs. `Symbol`) conflict. There are two ways to fix this:

1. Return `nil`.

```rb
# rb
class Foo < BasicObject
  def initialize
    @foo = :default
    nil
  end
end
```

2. Define the type for `Foo#initialize`.

```rbs
# rbs
class Foo < BasicObject
  def initialize: () -> Symbol
end
```

I don’t find either solution practical. Therefore, I believe the return type of `initialize` should be `untyped` or `void`.

To more actively convey the intent that the return type should not be used, I specify `void`.

This idea has now been implemented as a Cop in RuboCop and is being actively used.

[RBS/Style/InitializeReturnType](https://github.com/ksss/rubocop-on-rbs/blob/main/docs/modules/ROOT/pages/cops_rbs_style.adoc#rbsstyleinitializereturntype)
